src/source/PeerConnection/PeerConnection.c

```diff
@@ -84,12 +85,20 @@ VOID onInboundPacket(UINT64 customData, PBYTE buff, UINT32 buffLen)
             CHK_STATUS(putSctpPacket(pKvsPeerConnection->pSctpSession, buff, signedBuffLen));
         }
 
-        if (pKvsPeerConnection->pSrtpSession == NULL) {
-            dtlsSessionIsInitFinished(pKvsPeerConnection->pDtlsSession, &isDtlsConnected);
-            if (isDtlsConnected) {
-                allocateSrtp(pKvsPeerConnection);
+        CHK_STATUS(dtlsSessionIsInitFinished(pKvsPeerConnection->pDtlsSession, &isDtlsConnected));
+        if (isDtlsConnected) {
+            MUTEX_LOCK(pKvsPeerConnection->peerConnectionObjLock);
+            locked = TRUE;
+            if (pKvsPeerConnection->pSrtpSession == NULL) {
```

Comment:
These `NULL` checks seem fragile.

Maybe we should have the struct exist, but then pass the keys into SRTP (and start SCTP) when we are actually ready?

Reply:
The NULL check is inside `isDtlsConnected` though, so we are doing sctp and srtp allocate when we are actually ready. Because both SRTP and SCTP are initialized by the allocate call (key passing, start sctp etc.), the only way to check whether SRTP and SCTP has started seems to be the null check. To avoid duplicate allocate calls, I put them inside the lock.

Reply:
Does the ICE Agent dispatch packets concurrently? A lot of the PeerConnection code is written to assume it will be processing one packet a time

Reply:
actually it doesnt dispatch packets concurrently. so we dont actually need the lock here.